File: apps/mobile/lib/features/dm/data/dm_dto.dart
import '../../../core/json_helpers.dart';

/// P6.9 — Scoped DM DTOs (workflow-bounded private 1:1 channels).

class DmCounterpartDto {
  const DmCounterpartDto({required this.id, required this.nickname});
  final String id;
  final String? nickname;

  factory DmCounterpartDto.fromJson(Map<String, dynamic> j) => DmCounterpartDto(
        id: asString(j, 'id'),
        nickname: asStringOrNull(j, 'nickname'),
      );
}

class DmChannelDto {
  const DmChannelDto({
    required this.id,
    required this.scope,
    required this.refId,
    required this.counterpart,
    required this.status,
    required this.lastMessageAt,
    required this.unread,
    required this.createdAt,
  });

  final String id;
  final String scope; // RECRUITMENT | CONTRIBUTION
  final String refId;
  final DmCounterpartDto counterpart;
  final String status; // OPEN | CLOSED
  final DateTime? lastMessageAt;
  final bool unread;
  final DateTime? createdAt;

  bool get isClosed => status == 'CLOSED';

  factory DmChannelDto.fromJson(Map<String, dynamic> j) => DmChannelDto(
        id: asString(j, 'id'),
        scope: asString(j, 'scope'),
        refId: asString(j, 'ref_id'),
        counterpart: DmCounterpartDto.fromJson(
          asMap(j, 'counterpart') ?? const <String, dynamic>{},
        ),
        status: asString(j, 'status', fallback: 'OPEN'),
        lastMessageAt: asDateTimeOrNull(j, 'last_message_at'),
        unread: asBool(j, 'unread'),
        createdAt: asDateTimeOrNull(j, 'created_at'),
      );
}

class DmChannelListDto {
  const DmChannelListDto({required this.items});
  final List<DmChannelDto> items;

  factory DmChannelListDto.fromJson(Map<String, dynamic> j) =>
      DmChannelListDto(items: asObjectList(j, 'items', DmChannelDto.fromJson));
}

class DmMessageDto {
  const DmMessageDto({
    required this.id,
    required this.channelId,
    required this.senderId,
    required this.body,
    required this.status,
    required this.mine,
    required this.createdAt,
  });

  final String id;
  final String channelId;
  final String senderId;
  final String body;
  final String status; // VISIBLE | HIDDEN
  final bool mine;
  final DateTime? createdAt;

  bool get isHidden => status == 'HIDDEN';

  factory DmMessageDto.fromJson(Map<String, dynamic> j) => DmMessageDto(
        id: asString(j, 'id'),
        channelId: asString(j, 'channel_id'),
        senderId: asString(j, 'sender_id'),
        body: asString(j, 'body'),
        status: asString(j, 'status', fallback: 'VISIBLE'),
        mine: asBool(j, 'mine'),
        createdAt: asDateTimeOrNull(j, 'created_at'),
      );
}

class DmMessageListDto {
  const DmMessageListDto({
    required this.items,
    required this.nextCursor,
    required this.channelStatus,
  });
  final List<DmMessageDto> items;
  final String? nextCursor;
  final String channelStatus; // OPEN | CLOSED

  bool get isClosed => channelStatus == 'CLOSED';

  factory DmMessageListDto.fromJson(Map<String, dynamic> j) => DmMessageListDto(
        items: asObjectList(j, 'items', DmMessageDto.fromJson),
        nextCursor: asStringOrNull(j, 'next_cursor'),
        channelStatus: asString(j, 'channel_status', fallback: 'OPEN'),
      );
}

File: apps/api/src/modules/dm/dto/dm.dto.ts
// P6.9 — Scoped DM DTOs.

export type DmScope = 'RECRUITMENT' | 'CONTRIBUTION';

export interface CreateDmChannelInput {
  scope: string;
  ref_id: string;
  /**
   * Required only when the caller is the party_b side of a RECRUITMENT
   * channel (the post author opening a thread with a specific
   * applicant). For the applicant side, and for both CONTRIBUTION
   * sides, the counterpart is derived from the workflow.
   */
  counterpart_id?: string;
}

export interface SendDmMessageInput {
  body: string;
}

export interface DmCounterpartDTO {
  id: string;
  nickname: string | null;
}

export interface DmChannelDTO {
  id: string;
  scope: string;
  ref_id: string;
  counterpart: DmCounterpartDTO;
  status: string; // OPEN | CLOSED
  last_message_at: string | null;
  unread: boolean;
  created_at: string;
}

export interface DmChannelListDTO {
  items: DmChannelDTO[];
}

export interface DmMessageDTO {
  id: string;
  channel_id: string;
  sender_id: string;
  body: string;
  status: string; // VISIBLE | HIDDEN
  mine: boolean;
  created_at: string;
}

export interface DmMessageListDTO {
  items: DmMessageDTO[];
  next_cursor: string | null;
  channel_status: string;
}
